### ISSUE TITLE
Any plans to integrate this with Bevy?

### ISSUE THREAD

**Issue by thedodd** [alexxbb/hapi-rs#18]
Having a streamlined Houdini -> Bevy content creation workflow could be a bit of a game changer for building games/apps with Bevy.

**Comment by mike-mortimer** [comment]
Agreed, this kind of DCC pipeline would help Bevy adoption too.

**Comment by alexxbb** [comment]
Yes, it would be great to integrate it with Bevy. As a matter of fact I tried this about a year ago, but I didn't have much experience with Bevy at that time. ( And still don't have it 😆)

The biggest downside is licensing. You need a commercial Houdini license for third-party implementations like this one which can put off potential users. 

Having said that, I would like to learn Bevy better and try again this Fall.

**Comment by thedodd** [comment]
I'm going to be experimenting with this a bit for some stuff I am building. I'll keep it modular so that if I make progress on this, I'll be able to push the code to crates.io and such. Likely I'll use Bevy UI (for any UI stuff, params and such) to try to make it future proof.

My primary objective will be to use a Houdini `.hda` as a core part of a Bevy scene. That way I can do as much of the asset creation pipeline in Houdini, and then just load it as a Bevy scene.

One thing which I think will be quite important, and also quite doable, is to be able to export various attributes from Houdini, and have those attributes treated as Bevy components. This will be a nice way to integrate the asset creation pipeline with an abstract representation of logic. Then on the Bevy side, we just parse those attributes and attach them as components to the spawned entities using Bevy's reflection system (already a tried and true pattern, just not in the context of Houdini).

That would be killer. Once the Bevy editor ships, Bevy will be much closer to being a viable alternative to Unity and such. Well, it already is, but this will alleviate much of the boilerplate.

**Comment by alexxbb** [comment]
Hey @thedodd sorry for the late reply.

I'm curious to see what this may turn into. It can be a great test for the library - some APIs are missing (less common) and some design decisions were made to make it simpler and user friendly at the cost of maybe some overhead. But I'm open to changing it if necessary. 

I should also have more time to experiment with it later this year. 

Good luck!

**Comment by thedodd** [comment]
Well, unfortunately I only learned after the fact that the Indie license only grants access to the Unity & Unreal engines. In order to use the engine API for this use case, I would need to drop a pretty solid amount of money ... in order for me to do a bunch of work hahaha.

So, at this point, it is a no-op for me.

**Comment by alexxbb** [comment]
Yes, exactly, this is very unfortunate, but makes sense from SideFx business point of view. 

If Bevy (or Rust game dev in general) takes off and people start making money with it, then maybe it will be of SideFx interest to support another ecosystem.

Until then, this library is for niche use cases in VFX studios which have tons of Houdini licenses 🙂

**Comment by NicTanghe** [comment]
> Well, unfortunately I only learned after the fact that the Indie license only grants access to the Unity & Unreal engines. In order to use the engine API for this use case, I would need to drop a pretty solid amount of money ... in order for me to do a bunch of work hahaha.
> 
> So, at this point, it is a no-op for me.
@thedodd 

I`m pretty sure what you are saying is untrue.

The indie license gives you access to houdini engine wich is what is used by the engines to process the hda`s.

Im pretty sure about this but do correct me if I`m wrong

**Comment by alexxbb** [comment]
I'm pretty sure you can't use an Indie license with custom plugins which this library qualifies as.

Check out [this FAQ](https://www.sidefx.com/faq/houdini-engine-faq/#faq80)

I would happy to be wrong though.

**Comment by NicTanghe** [comment]
That`s about the Free apprentice license not the indie license.

Houdini indie ships with the aforementioned Houdini Engine license.
You can c that here
https://www.sidefx.com/products/compare/

they just have a different file extention

Op wo 2 okt 2024 om 00:43 schreef alexxbb ***@***.***>:

**Comment by NicTanghe** [comment]
That`s about the Free apprentice license not the indie license.

Houdini indie ships with the aforementioned Houdini Engine license.
You can c that here
https://www.sidefx.com/products/compare/

they just have a different file extention
you can find them at the bottem of the page

also a bit further down on the page you linked

License 	Cost 	Asset Types 	Houdini GUI 	UE5 Plug-in 	Unity Plug-in 	Maya/Max Plug-in 	Batch Processing
Houdini Engine for Unity 	FREE 	.HDA 	x 	YES 	YES 	x 	x
Houdini Engine for UE5 	FREE 	.HDA 	x 	YES 	YES 	x 	x
Houdini Engine 	$$ 	.HDA 	x 	YES 	YES 	YES 	YES
Houdini Engine Indie 	FREE 	.HDALC/.HDA 	x 	YES 	YES 	YES 	YES

**Comment by luxalpa** [comment]
I am using this with an indie license just fine.

**Comment by luxalpa** [comment]
I don't see why a higher tier license would be needed? I could perfectly integrate my Houdini Engine plugin with Bevy instead of using it in my custom game engine.

**Comment by doppioslash** [comment]
There are 2 key distinctions here:
1. Apprentice licenses vs Indie licenses vs Studio licenses
2. Houdini Engine for Unity&Unreal only licenses vs generic Houdini Engine license that will allow it to work with custom engines

With just an Houdini Apprentice license afaik you can't use Houdini Engine at all.

If you qualify for Houdini Indie licenses, you can have up to 3 *free* generic Houdini Engine licenses, that will work with any app, including a custom engine like Bevy.

If you need a Studio license, you only get free Houdini Engine for Unreal&Unity licenses, but not the generic Houdini Engine licenses.

Since Bevy is open source, I'd imagine if you write to SideFX they could be willing to help sort this out.

**Comment by alexxbb** [comment]
@NicTanghe yes, for this kind of integration SideFx would need to issue a special license. The issue is that no one can guarantee that this license won't be used by someone for commercial purposes. With the official Unity/Unreal plugins the license is distributed in binaries. SideFx would have to distribute the Bevy plugin themselves in a binary form and Im sure they are not interested in this at all with the current state of GameDev in Rust/Bevy. There are dozens of much more widely used game engines around which SideFx doesn't care about - they don't make money.

**Comment by NicTanghe** [comment]
@alexxbb 

Ok so what you`re saying is that if they issue a free license for usage to make a plugin for bevy.

they cant check if they would also make a plugin for developing a plugin for an in-house engine ?

I think they can check if someone doing business with said person suddenly has a Houdini plugin in their engine and they could require telemetry data to be on for usage. 
Although it would always be in good faith. and i be leave that sideFX has enough heart left for such a move. 

But they might get a relations hit with unreal/unity of they do decide to do something like this.

PS.
I think i might have never send a mail but I'm not sure and it will likely be a while before i try again..

PSS.
more people making games with a game that has Houdini support the more potential customers they have. Its a good way to take a part of blenders market. And if possible a godot engine should also be made.

**Comment by luxalpa** [comment]
> If you qualify for Houdini Indie licenses, you can have up to 3 free generic Houdini Engine licenses, that will work with any app, including a custom engine like Bevy.

It seems that this got changed stealthily this week. My Houdini Engine integration into my game no longer works with the error "Houdini Engine Indie license found but it can only be used with approved Houdini Engine plugins.".

**Comment by NicTanghe** [comment]
Yes well its hard to help the oligopolies if you don`t enforce your evil shemes.

**Comment by alexxbb** [comment]
Oh wow, could be that SideFx inadvertently allowed Indie license in third-party plugins without even realizing it!?

**Comment by luxalpa** [comment]
Yes, I have contacted SideFX support with this regard and they said using Indie Engine license for anything other than the official plugins was never supported. I also submitted an RFE and gave a bunch of reasons for why Houdini engine is very useful for indie devs.

**Comment by NicTanghe** [comment]
Great i whould second that RFE if i could but you need a studio license for that aswell.
I`m thinking it being usefull for indie devs is why it costs extra.

**Comment by alexxbb** [comment]
Hey just a heads up. I had some time to play around with Bevy over the Christmas holidays and made a simple Bevy app that can load an HDA and play geometry animation. 

Check out `apps/bevy` folder.
It's in the `main` branch and uses the latest `0.11.0` crate version which is **for Houdini 20.5**.
I don't have any immediate plans on Bevy plugin ATM, but we'll see.
